File: ascii_fields/animations/rain.py
import math
import random

from ..core import Animation, clamp, render_glyph_field


GLYPHS = "abcdefghijklmnopqrstuvwxyz0123456789@#$%&*+=<>|/\\:;.-"


class RainAnimation(Animation):
  """Matrix-style digital rain: columns of glyphs streaming downward with a
  bright leading head and a fading green-grey trail."""

  default_theme = "aurora"   # looks great in green truecolor; mono by default

  def __init__(self):
    self._w = 0
    self._h = 0
    self._heads = None
    self._speeds = None
    self._lengths = None
    self._rng = random.Random()

  def _seed(self, w, h):
    self._w, self._h = w, h
    self._offsets = [self._rng.uniform(0, h) for _ in range(w)]
    self._speeds = [self._rng.uniform(6.0, 18.0) for _ in range(w)]
    self._lengths = [self._rng.randint(5, max(6, h // 2)) for _ in range(w)]

  def render(self, width, height, elapsed, phase, options):
    if self._heads is None or width != self._w or height != self._h:
      self._heads = True
      self._seed(width, height)
    speed_mul = max(0.4, options.scale)
    heads = [(self._offsets[c] + elapsed * self._speeds[c] * speed_mul)
             % (height + self._lengths[c]) for c in range(width)]
    flick = int(elapsed * 12)
    grid = []
    glyphs = []
    for row in range(height):
      level_row = []
      glyph_row = []
      for col in range(width):
        head = heads[col]
        d = head - row
        length = self._lengths[col]
        if 0.0 <= d < length:
          if d < 1.0:
            level = 1.0
          else:
            level = clamp((1.0 - d / length) * 0.85)
          gi = (col * 31 + row * 17 + (flick if d < 2 else flick // 6)) % len(GLYPHS)
          glyph_row.append(GLYPHS[gi])
          level_row.append(level)
        else:
          glyph_row.append(" ")
          level_row.append(0.0)
      grid.append(level_row)
      glyphs.append(glyph_row)
    return render_glyph_field(width, height, grid, glyphs, options, self)

File: src/animations/rain.rs
use rand::{Rng, SeedableRng};
use rand_pcg::Pcg32;
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_glyph_field, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "aurora" };
const GLYPHS: &str = "abcdefghijklmnopqrstuvwxyz0123456789@#$%&*+=<>|/\\:;.-";

pub struct Rain {
  w: usize, h: usize,
  offsets: Vec<f64>,
  speeds: Vec<f64>,
  lengths: Vec<i64>,
  rng: Pcg32,
}
impl Default for Rain {
  fn default() -> Self {
    Self { w: 0, h: 0, offsets: Vec::new(), speeds: Vec::new(), lengths: Vec::new(), rng: Pcg32::from_entropy() }
  }
}

impl Rain {
  fn seed(&mut self, w: usize, h: usize) {
    self.w = w; self.h = h;
    self.offsets = (0..w).map(|_| self.rng.gen_range(0.0..h as f64)).collect();
    self.speeds = (0..w).map(|_| self.rng.gen_range(6.0..18.0)).collect();
    let lo = 5_i64; let hi = (6_i64).max(h as i64 / 2);
    self.lengths = (0..w).map(|_| self.rng.gen_range(lo..=hi)).collect();
  }
}

impl Animation for Rain {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    if self.offsets.is_empty() || ctx.width != self.w || ctx.height != self.h {
      self.seed(ctx.width, ctx.height);
    }
    let speed_mul = ctx.options.scale.max(0.4);
    let w = ctx.width; let h = ctx.height;
    let chars: Vec<char> = GLYPHS.chars().collect();
    let n = chars.len();
    let mut heads = vec![0.0_f64; w];
    for c in 0..w {
      heads[c] = ((self.offsets[c] + ctx.elapsed * self.speeds[c] * speed_mul)
                  % (h as f64 + self.lengths[c] as f64) + h as f64 + self.lengths[c] as f64)
                  % (h as f64 + self.lengths[c] as f64);
    }
    let flick = (ctx.elapsed * 12.0) as i64;
    let mut grid = vec![0.0_f64; w * h];
    let mut glyphs = vec![' '; w * h];
    for row in 0..h {
      for col in 0..w {
        let head = heads[col];
        let d = head - row as f64;
        let length = self.lengths[col];
        let (level, glyph) = if d >= 0.0 && d < length as f64 {
          let level = if d < 1.0 { 1.0 } else { clamp((1.0 - d / length as f64) * 0.85) };
          let f = if d < 2.0 { flick } else { flick / 6 };
          let gi = (((col * 31 + row * 17) as i64 + f) % n as i64).rem_euclid(n as i64) as usize;
          (level, chars[gi])
        } else { (0.0, ' ') };
        grid[row * w + col] = level;
        glyphs[row * w + col] = glyph;
      }
    }
    render_glyph_field(ctx, &grid, &glyphs, &STYLE, out);
  }
}
